python/pyiceberg/expressions/visitors.py

```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_greater_than_or_equal(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        field = term.ref().field
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.upper_bounds and from_bytes(field.field_type, self.upper_bounds[idx]) < literal.value:
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_equal(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        field = term.ref().field
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.lower_bounds:
+            lower = from_bytes(field.field_type, self.lower_bounds[idx])
+
+            if self._is_nan(lower):
+                # NaN indicates unreliable bounds. See the InclusiveMetricsEvaluator docs for more.
+                return ROWS_MIGHT_MATCH
+
+            if lower > literal.value:
+                return ROWS_CANNOT_MATCH
+
+        if idx in self.upper_bounds and from_bytes(field.field_type, self.upper_bounds[idx]) < literal.value:
+            return ROWS_CANNOT_MATCH
+
+        return ROWS_MIGHT_MATCH
+
+    def visit_not_equal(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        return ROWS_MIGHT_MATCH
+
+    def visit_in(self, term: BoundTerm[L], literals: Set[L]) -> bool:
+        field = term.ref().field
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if len(literals) > IN_PREDICATE_LIMIT:
+            # skip evaluating the predicate if the number of values is too big
+            return ROWS_MIGHT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.lower_bounds:
+            lower = from_bytes(field.field_type, self.lower_bounds[idx])
+
+            if self._is_nan(lower):
+                # NaN indicates unreliable bounds. See the InclusiveMetricsEvaluator docs for more.
+                return ROWS_MIGHT_MATCH
+
+            literals = {lit for lit in literals if lower <= lit}
+            if len(literals) == 0:
+                return ROWS_CANNOT_MATCH
+
+        if idx in self.upper_bounds:
+            upper = from_bytes(field.field_type, self.upper_bounds[idx])
+            # this is different from Java, here NaN is always larger
+            if not self._is_nan(upper):
```

Comment:
This seems equivalent to the `if self._is_nan(lower)` in other places. Maybe we should use the same check and return `ROWS_MIGHT_MATCH` if the bound is NaN.

Reply:
Good one, updated 👍🏻

Reply:
Looks like this was updated for `visit_in` but not `visit_equal`

Reply:
Missed those, added!